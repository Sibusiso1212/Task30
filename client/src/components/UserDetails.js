import React, { useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';
import '../styles/UserDetails.css';

// Component to display detailed information about a GitHub user and their repositories
const UserDetails = () => {
    // Extract username from URL parameters
    const { username } = useParams();
    // State for storing user and repository data
    const [userData, setUserData] = useState(null);
    // State for managing loading state
    const [loading, setLoading] = useState(true);

    // Effect hook to fetch user data when component mounts or username changes
    useEffect(() => {
        // Async function to fetch user data from the backend
        const fetchUserData = async () => {
            try {
                const response = await fetch(`http://localhost:5000/api/users/${username}`);
                const data = await response.json();
                setUserData(data);
            } catch (error) {
                console.error(error);
            }
            setLoading(false);
        };

        fetchUserData();
    }, [username]);

    // Helper function to format dates in a readable format
    const formatDate = (dateString) => {
        return new Date(dateString).toLocaleDateString('en-US', {
            year: 'numeric',
            month: 'long',
            day: 'numeric'
        });
    };

    // Show loading state while data is being fetched
    if (loading) return <div className="loader">Loading...</div>;
    // Show error message if user data is not found
    if (!userData) return <div>User not found</div>;

    return (
        <div className="user-details">
            {/* User Profile Section */}
            <div className="profile-section">
                <img src={userData.user.avatar_url} alt={userData.user.login} />
                <h2>{userData.user.name || userData.user.login}</h2>
                <p className="bio">{userData.user.bio}</p>
                {/* User Statistics */}
                <div className="stats">
                    <span>Followers: {userData.user.followers}</span>
                    <span>Following: {userData.user.following}</span>
                    <span>Public Repos: {userData.user.public_repos}</span>
                </div>
                <a href={userData.user.html_url} target="_blank" rel="noopener noreferrer" className="external-link">
                    View on GitHub
                </a>
            </div>
            
            {/* Repositories Section */}
            <div className="repos-section">
                <h3>Repositories</h3>
                <div className="repos-grid">
                    {/* Map through repositories and display them as cards */}
                    {userData.repos.map(repo => (
                        <div key={repo.id} className="repo-card">
                            <h4>{repo.name}</h4>
                            <p className="description">{repo.description || 'No description available'}</p>
                            {/* Repository Details */}
                            <div className="repo-details">
                                <p><strong>Created:</strong> {formatDate(repo.created_at)}</p>
                                <p><strong>Last Updated:</strong> {formatDate(repo.updated_at)}</p>
                                <p><strong>Last Push:</strong> {formatDate(repo.pushed_at)}</p>
                                <p><strong>Language:</strong> {repo.language || 'Not specified'}</p>
                                <p><strong>Stars:</strong> {repo.stargazers_count}</p>
                                <p><strong>Forks:</strong> {repo.forks_count}</p>
                            </div>
                            <a href={repo.html_url} target="_blank" rel="noopener noreferrer" className="repo-link">
                                View Repository
                            </a>
                        </div>
                    ))}
                </div>
            </div>
        </div>
    );
};

export default UserDetails;