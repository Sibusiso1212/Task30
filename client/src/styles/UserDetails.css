.user-details {
  max-width: 1200px;
  margin: 0 auto;
  padding: 2rem;
}

.profile-section {
  background: white;
  border-radius: 20px;
  padding: 2rem;
  text-align: center;
  box-shadow: var(--card-shadow);
  margin-bottom: 2rem;
}

.profile-section img {
  width: 180px;
  height: 180px;
  border-radius: 50%;
  border: 6px solid var(--secondary-color);
  margin-bottom: 1.5rem;
}

.external-link {
  display: inline-block;
  padding: 0.75rem 1.5rem;
  background: var(--accent-color);
  color: white;
  text-decoration: none;
  border-radius: 10px;
  margin-top: 1rem;
  transition: all 0.3s ease;
}

.external-link:hover {
  background: #6d28d9;
  transform: translateY(-2px);
}

.repos-section {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(300px, 1fr));
  gap: 2rem;
}
.repo-card {
  background: white;
  border-radius: 16px;
  padding: 1.5rem;
  box-shadow: 0 4px 6px -1px rgba(0, 0, 0, 0.1);
  transition: transform 0.3s ease;
}

.repo-card:hover {
transform: translateY(-3px);
}

.repo-card h4 {
  color: #2d333b;
  font-size: 1.25rem;
  margin-bottom: 1rem;
}

.repo-card .description {
  color: #64748b;
  margin-bottom: 1rem;
  line-height: 1.5;
}

.repo-details {
  background: #f8fafc;
  padding: 1rem;
  border-radius: 8px;
  margin: 1rem 0;
}

.repo-details p {
  margin: 0.5rem 0;
  color: #475569;
}

.repo-details strong {
  color: #2d333b;
}

.repo-link {
  display: inline-block;
  padding: 0.75rem 1.5rem;
  background: #7c3aed;
  color: white;
  text-decoration: none;
  border-radius: 8px;
  transition: background-color 0.3s ease;
}

.repo-link:hover {
  background: #6d28d9;
}

.stats {
  display: flex;
  gap: 2rem;
  justify-content: center;
  margin: 1rem 0;
}

.stats span {
  color: #64748b;
  font-size: 0.9rem;
}
.loader {
  display: flex;
  justify-content: center;
  align-items: center;
  min-height: 200px;
  font-size: 1.5rem;
  color: var(--accent-color);
}
