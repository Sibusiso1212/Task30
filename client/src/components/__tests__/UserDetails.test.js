import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import UserDetails from '../UserDetails';

// Mock fetch function
global.fetch = jest.fn();

describe('UserDetails Component', () => {
    // Snapshot test
    test('matches snapshot', () => {
        const { container } = render(
            <MemoryRouter initialEntries={['/user/testuser']}>
                <Routes>
                    <Route path="/user/:username" element={<UserDetails />} />
                </Routes>
            </MemoryRouter>
        );
        expect(container).toMatchSnapshot();
    });

    // Loading state test
    test('displays loading state', () => {
        render(
            <MemoryRouter initialEntries={['/user/testuser']}>
                <Routes>
                    <Route path="/user/:username" element={<UserDetails />} />
                </Routes>
            </MemoryRouter>
        );
        expect(screen.getByText('Loading...')).toBeInTheDocument();
    });

    // User data display test
    test('displays user data when loaded', async () => {
        const mockData = {
            user: {
                login: 'testuser',
                name: 'Test User',
                bio: 'Test Bio',
                avatar_url: 'test.jpg'
            },
            repos: []
        };

        global.fetch.mockResolvedValueOnce({
            json: () => Promise.resolve(mockData)
        });

        render(
            <MemoryRouter initialEntries={['/user/testuser']}>
                <Routes>
                    <Route path="/user/:username" element={<UserDetails />} />
                </Routes>
            </MemoryRouter>
        );

        await waitFor(() => {
            expect(screen.getByText('Test User')).toBeInTheDocument();
            expect(screen.getByText('Test Bio')).toBeInTheDocument();
        });
    });
});
