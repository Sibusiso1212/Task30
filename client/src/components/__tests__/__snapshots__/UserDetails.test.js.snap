// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDetails Component matches snapshot 1`] = `
<div>
  <div
    class="loader"
  >
    Loading...
  </div>
</div>
`;
